### PR TITLE
Implement Store::retrieve

### PR DESCRIPTION
**Pull request by TheNeikos**

### CODE REVIEW

**Comment by neithernut** [comment]
Mh... I think we'll end up with three different local names for the hashmap containing the entries

File: libimagstore/src/store.rs [hunk 1]
```diff
@@ -122,7 +122,17 @@ impl Store {
...
+        if hsmap.is_err() {
+            return Err(StoreError::new(StoreErrorKind::LockPoisoned, None))
+        }
+        hsmap.unwrap().get_mut(&id)
+            .ok_or(StoreError::new(StoreErrorKind::IdNotFound, None))
+            .and_then(|store_entry| {
+                store_entry.get_entry()
+            }).and_then(|entry| {
+                Ok(FileLockEntry::new(self, entry, id))
+            })
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: You can put the `and_then()` into single lines each.

File: libimagstore/src/store.rs [hunk 2]
```diff
@@ -122,7 +122,17 @@ impl Store {
...
-        unimplemented!();
+        let hsmap = self.entries.write();
+        if hsmap.is_err() {
+            return Err(StoreError::new(StoreErrorKind::LockPoisoned, None))
+        }
+        hsmap.unwrap().get_mut(&id)
+            .ok_or(StoreError::new(StoreErrorKind::IdNotFound, None))
+            .and_then(|store_entry| {
+                store_entry.get_entry()
+            }).and_then(|entry| {
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: If you don't want to put them `and_then()` calls into single lines , please put this into a new line at least:

```rs
.and_then(|s| {
})
.and_then(|x| {
})
```

Comment (neithernut) [inline libimagstore/src/store.rs]:
```
.and_then(|store_entry| store_entry.get_entry())
.and_then(...)
```
would also be fine in this case, wouldn't it?

Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Yes, totally. As long as you don't go over 100 characters per line.

For all these `and_then`, `or_else`, `map` and so on, it is basically:

* If you call one function, you don't need a block (`{ ... }`) most of the time
* If you _need_ a block, do it mutliline

and my opinion would be to _always_ start lines with a dot, if you have to chain up some of these calls (like we both just showed).

Comment (neithernut) [inline libimagstore/src/store.rs]: Ah, I removed the comment again because I thought it was redundant.

Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: nvm.

File: libimagstore/src/store.rs [hunk 3]
```diff
@@ -122,7 +122,17 @@ impl Store {
...
     pub fn retrieve<'a>(&'a self, id: StoreId) -> Result<FileLockEntry<'a>> {
-        unimplemented!();
+        let hsmap = self.entries.write();
+        if hsmap.is_err() {
+            return Err(StoreError::new(StoreErrorKind::LockPoisoned, None))
+        }
+        hsmap.unwrap().get_mut(&id)
+            .ok_or(StoreError::new(StoreErrorKind::IdNotFound, None))
+            .and_then(|store_entry| {
+                store_entry.get_entry()
```
Comment (neithernut) [inline libimagstore/src/store.rs]: Now that `get_entry()` is gone, this won't work. You'll have to read the entry using the new mechanism.

Comment (neithernut) [inline libimagstore/src/store.rs]: Oh, we don't have a new mechanism yet...
I'll open an issue.

File: libimagstore/src/store.rs [hunk 4]
```diff
@@ -24,16 +29,34 @@ enum StoreEntryPresence {
...
 impl StoreEntry {
     /// The entry is currently borrowed, meaning that some thread is currently
     /// mutating it
     fn is_borrowed(&self) -> bool {
-        self.entry == StoreEntryPresence::Borrowed
+        self.status == StoreEntryStatus::Borrowed
+    }
+
+    fn get_entry(&mut self) -> Result<Entry> {
+        if self.is_borrowed() {
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: `!` missing.

File: libimagstore/src/store.rs (hunk 3 above)
Comment (TheNeikos) [inline libimagstore/src/store.rs]: It has been implemented

**Comment by neithernut** [comment]
Mh... I don't want to annoy you, but please have a look at #139.

**Comment by TheNeikos** [comment]
Well yeah, still WIP :dancer:

**Comment by TheNeikos** [comment]
Ready for Review

File: libimagstore/src/store.rs [hunk 5]
```diff
@@ -83,7 +106,14 @@ impl Store {
...
     /// dropped, the new Entry is written to disk
     pub fn retrieve<'a>(&'a self, id: StoreId) -> Result<FileLockEntry<'a>> {
-        unimplemented!();
+        let hsmap = self.entries.write();
+        if hsmap.is_err() {
+            return Err(StoreError::new(StoreErrorKind::LockPoisoned, None))
+        }
+        hsmap.unwrap().get_mut(&id)
+            .ok_or(StoreError::new(StoreErrorKind::IdNotFound, None))
+            .and_then(|store_entry| store_entry.get_entry() )
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Pls remove the whitespace in `get_entry() )` (before the last closing `)`)

File: libimagstore/src/store.rs [hunk 6]
```diff
@@ -127,11 +157,9 @@ impl Drop for Store {
     /**
      * Unlock all files on drop
      *
-     * TODO: Error message when file cannot be unlocked?
+     * TODO: Unlock them
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Will this be done in another PR?

File: libimagstore/src/store.rs [hunk 7]
```diff
@@ -172,3 +200,346 @@ impl<'a> Drop for FileLockEntry<'a> {
...
     }
 }
+
+/**
+ * EntryContent type
+ */
+pub type EntryContent = String;
+
+use error::ParserErrorKind;
+use error::ParserError;
```
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Why not placing these at the top of the file?

File: libimagstore/src/store.rs (hunk 6 above)
Comment (TheNeikos) [inline libimagstore/src/store.rs]: I could do it here if you wish, it's just not very pertinent to retrieve

**Comment by matthiasbeyer** [comment]
I guess this looks good. Fell free to cleanup the history.

**Comment by matthiasbeyer** [comment]
Waiting on CI.

File: libimagstore/src/store.rs (hunk 6 above)
Comment (matthiasbeyer) [inline libimagstore/src/store.rs]: Implement it in another PR, so we get this one into the codebase!
